TESTS/mbed_drivers/unbuffered_Serial/gt_unbuffered_serial_test_env.cpp

```diff
@@ -0,0 +1,341 @@
+}
+
+
+/**
+ *  \brief Key-value parser
+ *
+ *         Key-value message grammar
+ *
+ *         <MESSAGE>: <TOK_OPEN> <TOK_STRING> <TOK_SEMICOLON> <TOK_STRING> <TOK_CLOSE>
+ *
+ *         Examples:
+ *         message:     "{{__timeout; 1000}}"
+ *                      "{{__sync; 12345678-1234-5678-1234-567812345678}}"
+ *
+ *  \param out_key Output buffer to store key string value
+ *  \param out_value Output buffer to store value string value
+ *  \param out_key_size Buffer 'out_key' buffer size
+ *  \param out_value_size Buffer 'out_value_size' buffer size
+ *  \return Returns 1 if key-value message was parsed successfully in stream of tokens from tokenizer
+ *
+ */
+static int HandleKV(char *out_key,
+                    char *out_value,
+                    const int out_key_size,
+                    const int out_value_size) {
+    // We already started with <open>
+    if (getNextToken(out_key, out_key_size) == tok_string) {
+        if (getNextToken(0, 0) == tok_semicolon) {
+            if (getNextToken(out_value, out_value_size) == tok_string) {
+                if (getNextToken(0, 0) == tok_close) {
+                    // <open> <string> <semicolon> <string> <close>
+                    // Found "{{KEY;VALUE}}" expression
+                    return 1;
+                }
+            }
+        }
+    }
+    getNextToken(0, 0);
+    return 0;
+}
+
+/**
+ * \brief parse input string for key-value pairs: {{key;value}}
+ *        This function should replace scanf() used to
+ *        check for incoming messages from master. All data
+ *        parsed and rejected is discarded.
+ *
+ * \param out_key Ouput data with key
+ * \param out_value Ouput data with value
+ * \param out_key_size out_key total size
+ * \param out_value_size out_value total data
+ *
+ * success != 0 when key-value pair was found
+ * success == 0 when end of the stream was found
+ *
+ */
+int greentea_unbuffered_serial_parse_kv(
+    char *out_key,
+    char *out_value,
+    const int out_key_size,
+    const int out_value_size
+) {
+    getNextToken(0, 0);
+    while (1) {
+        switch (cur_token) {
+        case tok_eof:
+            return 0;
+
+        case tok_open:
+            if (HandleKV(out_key, out_value, out_key_size, out_value_size)) {
+                // We've found {{ KEY ; VALUE }} expression
+                return 1;
+            }
+            break;
+
+        default:
+            // Load next token and pray...
+            getNextToken(0, 0);
+            break;
+        }
+    }
+}
+
+/**
+ * \brief Encapsulate and send key-value message from DUT to host
+ *
+ *        This function uses underlying functions to write directly
+ *        to the serial port, (USBTX). This allows KVs to be used
+ *        from within interrupt context.
+ *
+ * \param key Message key (message/event name)
+ * \param value Message payload, string value
+ * 
+ * \return Number of bytes sent
+ *
+ */
+ssize_t greentea_unbuffered_serial_send_kv(
```

Comment:
How are you gonna call `UnbufferedSerial::write` without any message formatting? The message needs to be formatted in a way that Greentea understands it before `UnbufferedSerial::write` is called and that is the purpose of the function.

Reply:
I would write:
```
char message[] = "{{MSG_KEY_ECHO_MESSAGE, MESSAGE2}}";
test_unbuffered_serial.write(message, sizeof(message));
```

You need to pass the entire string (formatted as required because of the greentea protocol) to `write` so the test can compare it in step 3 below
**Write test steps:**
1. Target writes data which turns out to be {{MSG_KEY_ECHO_MESSAGE, MESSAGE2}}
2. Host sends back {{MSG_KEY_ECHO_MESSAGE, MESSAGE2}}
3. Test asserts that data read by the host is what was expected.

Reply:
Please review [this](https://github.com/hugueskamba/mbed-os/compare/310a7cd73ede0bcfff6c99b4b3ae1c891754b989..dd9ce703013101cd3ddc946a7e0f393357532b49).